packages/mdc-data-table/foundation.ts

```diff
@@ -0,0 +1,193 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import {MDCFoundation} from '@material/base/foundation';
+import {MDCDataTableAdapter} from './adapter';
+import {cssClasses, strings} from './constants';
+
+export class MDCDataTableFoundation extends MDCFoundation<MDCDataTableAdapter> {
+  static get defaultAdapter(): MDCDataTableAdapter {
+    return {
+      addClassAtRowIndex: () => undefined,
+      getRowCount: () => 0,
+      getRowElements: () => [],
+      getRowIdAtIndex: () => '',
+      getRowIndexByChildElement: () => 0,
+      getSelectedRowCount: () => 0,
+      isCheckboxAtRowIndexChecked: () => false,
+      isHeaderRowCheckboxChecked: () => false,
+      isRowsSelectable: () => false,
+      notifyRowSelectionChanged: () => undefined,
+      notifySelectedAll: () => undefined,
+      notifyUnselectedAll: () => undefined,
+      registerHeaderRowCheckbox: () => undefined,
+      registerRowCheckboxes: () => undefined,
+      removeClassAtRowIndex: () => undefined,
+      setAttributeAtRowIndex: () => undefined,
+      setHeaderRowCheckboxChecked: () => undefined,
+      setHeaderRowCheckboxIndeterminate: () => undefined,
+      setRowCheckboxCheckedAtIndex: () => undefined,
+    } as MDCDataTableAdapter;
+  }
+
+  constructor(adapter?: Partial<MDCDataTableAdapter>) {
+    super({...MDCDataTableFoundation.defaultAdapter, ...adapter});
+  }
+
+  init() {
+    this.layout();
+  }
+
+  /**
+   * Re-initializes header row checkbox and row checkboxes when selectable rows are added or removed from table.
+   * Use this if registering checkbox is synchronous.
+   */
+  layout() {
+    if (this.adapter_.isRowsSelectable()) {
+      this.adapter_.registerHeaderRowCheckbox();
+      this.adapter_.registerRowCheckboxes();
+
+      this.setHeaderRowCheckboxState_();
+    }
+  }
+
+  /**
+   * Re-initializes header row checkbox and row checkboxes when selectable rows are added or removed from table.
+   * Use this if registering checkbox is asynchronous.
+   */
+  async layoutAsync(): Promise<void> {
```

Comment:
This doesn't return anything

Reply:
async/await implicitly returns `Promise` type.

Reply:
ah ok

Reply:
So technically this should be `Promise<void> | void`